src/js.c

```diff
@@ -2927,20 +2927,22 @@ js_maybeFetchMore(natsSubscription *sub, jsFetch *fetch)
     natsBuf_InitWithBackend(&buf, buffer, 0, sizeof(buffer));
 
     nats_lockSubAndDispatcher(sub);
-
     jsSub *jsi = sub->jsi;
     jsi->inFetch = true;
     jsi->fetchID++;
     snprintf(fetch->replySubject, sizeof(fetch->replySubject), "%.*s%" PRIu64,
              (int)strlen(sub->subject) - 1, sub->subject, // exclude the last '*'
              jsi->fetchID);
-    natsStatus s = _sendPullRequest(sub->conn, jsi->nxtMsgSubj, fetch->replySubject, &buf, &req);
+    nats_unlockSubAndDispatcher(sub);
+
+    natsStatus s = _publishPullRequest(sub->conn, jsi->nxtMsgSubj, fetch->replySubject, &buf, &req);
```

Comment:
Is there not a possibility that js_maybeFetchMore be called concurrently for the same `sub->jsi`? If so, then `fetch->replySubject` that is "set" above under the lock is accessed here without and may have changed in a concurrent call. Or is the `fetch` object guaranteed to be used from a single thread?

Reply:
@kozlovic with the exception of the initial call, all other calls were made from the sub's delivery thread, but I decided to move it into a locally malloced buffer anyway, seems cleaner and these requests are relatively infrequent.

The initial invocation was inside the (recursive) lock, I took it out of there too.